### PR TITLE
Tweak candidate selection

### PR DESCRIPTION
**Pull request by starcke**
Tweak candidate selection:
- Do not send methods modeled outside the file as candidates.
- At most 6 usages.
- Send -1 for `this` argument.
- Do not send only some arguments for candidate or sample.

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: extensions/ql-vscode/src/data-extensions-editor/auto-model.ts [hunk 1]
```diff
@@ -40,8 +46,10 @@ export function createAutoModelRequest(
         ? 0
         : externalApiUsage.methodParameters.split(",").length;
 
+    const candidates = [];
+    const samples = [];
```
Comment (koesie10) [inline extensions/ql-vscode/src/data-extensions-editor/auto-model.ts:50]:
```suggestion
    const candidates: Method[] = [];
    const samples: Method[] = [];
```

**Review by koesie10 — COMMENTED** [review]
My main concern with this PR is that we're not using the correct name for the `input`. In the most recent versions of CodeQL, the MaD input for this is `Argument[this]`, while we would be sending and saving `Argument[-1]`, which would not work.

Perhaps this should be dealt with in the pre-processing on the LLM side rather than in the extension? I believe methods that are being modeled as `Argument[-1]` will now simply not work when running queries with them.

File: extensions/ql-vscode/src/data-extensions-editor/auto-model.ts [hunk 2]
```diff
@@ -40,8 +46,10 @@ export function createAutoModelRequest(
...
     for (
-      let argumentIndex = 0;
+      let argumentIndex = -1; // Start at -1 which means `this` as in `this.method()`
```
[the remark this answers is not in the document]
Comment (starcke) [reply extensions/ql-vscode/src/data-extensions-editor/auto-model.ts:52]: Yeah I also thought about that. It probably fits into the discussion we had about what endpoints to model.

**Comment by starcke** [comment]
> My main concern with this PR is that we're not using the correct name for the input. In the most recent versions of CodeQL, the MaD input for this is Argument[this], while we would be sending and saving Argument[-1], which would not work.

Yeah I agree. My plan was to deal with it in the extension by mapping `-1` back to `Argument[this]`. However maybe that is a bit to fiddly and we can just send the right thing (and update turbomodel to use input instead of ArgumentIndex. That is also the way CodeML does in their newest version.

I can update the PR to send `Argument[this]` (still unconditionally) and then make a PR for turbomodel?

**Comment by koesie10** [comment]
> I can update the PR to send `Argument[this]` (still unconditionally) and then make a PR for turbomodel?

Sounds good, I think that's perfect, then the extension doesn't need to care about the implementation details of the LLM.

**Comment by starcke** [comment]
Made the change and have a PR for turbomodel here: https://github.com/github/turbomodel/pull/44

**Review by koesie10 — APPROVED** [review]

**Review by charisk — APPROVED** [review]

**Comment by starcke** [comment]
Merged https://github.com/github/turbomodel/pull/44 and verified that I got a sink classified as `Argument[this]` back.

**Pull request merged**